src/DSE.Open/Text/StringHelper.Join.cs

```diff
@@ -0,0 +1,505 @@
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+using System.Text;
+
+namespace DSE.Open.Text;
+
+public static partial class StringHelper
+{
+    /// <summary>
+    /// Concatenates the members of a constructed <see cref="IEnumerable{T}"/> collection of type <see cref="string"/>,
+    /// using the specified <paramref name="separator"/> between each member before the penultimate, and the
+    /// specified <paramref name="finalSeparator"/> between the penultimate and final members.
+    /// </summary>
+    /// <param name="separator">The string to use as a separator. It is included in the returned string only if
+    /// <paramref name="values"/> has more than one element.</param>
+    /// <param name="values">A collection that contains the strings to concatenate.</param>
+    /// <param name="finalSeparator">The string to use as the final separator. It is included in the returned string
+    /// only if <paramref name="values"/> has more than one element.</param>
+    /// <returns>A string that consists of the elements of <paramref name="values"/> delimited by the
+    /// <paramref name="separator"/> string between each member before the penultimate, and by the
+    /// <paramref name="finalSeparator"/> between the penultimate and final members; or, <see cref="string.Empty"/>
+    /// if <paramref name="values"/> has zero elements.</returns>   
+    public static string Join(string? separator, IEnumerable<string?> values, string? finalSeparator)
+    {
+        if (values is List<string?> valuesList)
+        {
+            return JoinCore(separator.AsSpan(), CollectionsMarshal.AsSpan(valuesList), finalSeparator.AsSpan());
+        }
+
+        if (values is string?[] valuesArray)
+        {
+            return JoinCore(separator.AsSpan(), new ReadOnlySpan<string?>(valuesArray), finalSeparator.AsSpan());
+        }
+
+        ArgumentNullException.ThrowIfNull(values);
+
+        if (values is ICollection<string?> collection)
+        {
+            var count = collection.Count;
+
+            if (count == 0)
+            {
+                return string.Empty;
+            }
+
+            var charCount = collection.Sum(s => s is not null ? (long)s.Length : 0)
+                            + ((count - 2) * separator?.Length ?? 0)
+                            + finalSeparator?.Length ?? 0;
+
+            if (charCount > int.MaxValue)
+            {
+                ThrowHelper.ThrowInvalidOperationException("Resulting string would be > Int32.MaxValue characters.");
+            }
+
+            var length = (int)charCount;
+
+            if (length == 0)
+            {
+                return string.Empty;
+            }
+
+            if (length <= StackallocThresholds.MaxCharLength)
+            {
+                return JoinPreAllocated(separator.AsSpan(), collection, finalSeparator.AsSpan(), length);
+            }
+
+            return string.Create(length, collection, (chars, state) =>
```

Comment:
This should take `count` in the state, otherwise it's going to allocate a new delegate.

Reply:
Good catch 👍